File: ios/GradeThread/Referrals/ReferralTypes.swift
import Foundation

// Referrals wire models. Mirror the web contract (`src/pages/referrals.tsx`) and
// the edge route `services/edge-functions/src/routes/referrals.ts`. Decoded via
// `EdgeAPI`'s shared decoder (snake_case → camelCase: `referred_by` → `referredBy`).

/// `GET /api/referrals/me`
struct ReferralMe: Decodable, Equatable {
    let code: String
    let stats: ReferralStats
    let referredBy: ReferredBy?
}

struct ReferralStats: Decodable, Equatable {
    let total: Int
    let pending: Int
    let qualified: Int
    let granted: Int
}

struct ReferredBy: Decodable, Equatable {
    let status: String
    let code: String
}

/// `POST /api/referrals/redeem`
///
/// On success the edge returns `{ ok: true }`. On a business rejection (invalid /
/// self / already-referred / suspended code) the edge returns a 4xx whose body
/// carries a machine-readable `error_code`; `ReferralService` folds that into
/// `RedeemResponse(ok: false, reason:)` so the store can map it to specific copy
/// (US-1255) instead of one generic "that code isn't valid".
struct RedeemResponse: Decodable, Equatable {
    let ok: Bool
    /// Machine-readable rejection reason (`error_code`) when `ok == false`.
    let reason: String?

    init(ok: Bool, reason: String? = nil) {
        self.ok = ok
        self.reason = reason
    }
}

/// Maps the edge's redeem rejection `error_code` to specific, friendly copy.
/// Pure + unit-tested (US-1255) so the reason→message contract is verifiable
/// without the network. An unknown or absent reason falls back to `generic`.
enum RedeemRejection {
    static let generic = "That code isn't valid. Double-check it and try again."

    static func message(for reason: String?) -> String {
        switch reason {
        case "invalid_code", "not_found":
            return "We couldn't find that code. Double-check it and try again."
        case "self_referral":
            return "That's your own code — enter a friend's code instead."
        case "already_referred":
            return "You've already applied a referral code."
        case "expired":
            return "That code has expired. Ask your friend for a current one."
        case "account_suspended":
            return "Your account can't redeem a referral code right now."
        case "missing_code":
            return "Enter a code to continue."
        default:
            return generic
        }
    }
}

File: android/app/src/main/java/com/gradethread/app/referrals/ReferralTypes.kt
package com.gradethread.app.referrals

import androidx.annotation.StringRes
import com.gradethread.app.R

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * US-1385 (iOS `ReferralTypes`): the referrals wire contract.
 *
 * Mirrors `services/edge-functions/src/routes/referrals.ts` and the web's
 * `src/pages/referrals.tsx`. Every field is defaulted, because the edge grew
 * `credits` and `milestones` after the first clients shipped and a strict decode
 * would break an old build against a new server for no reason.
 */
@Serializable
data class ReferralMe(
    val code: String = "",
    val stats: ReferralStats = ReferralStats(),
    val credits: ReferralCredits = ReferralCredits(),
    val milestones: ReferralMilestones = ReferralMilestones(),
    @SerialName("referred_by") val referredBy: ReferredBy? = null,
)

@Serializable
data class ReferralStats(val total: Int = 0, val pending: Int = 0, val qualified: Int = 0, val granted: Int = 0)

@Serializable
data class ReferralCredits(
    @SerialName("per_referral") val perReferral: Int = 0,
    val earned: Int = 0,
    val pending: Int = 0,
)

@Serializable
data class ReferralMilestones(
    @SerialName("earned_bonus_credits") val earnedBonusCredits: Int = 0,
    val next: NextMilestone? = null,
)

@Serializable
data class NextMilestone(val threshold: Int = 0, val bonus: Int = 0, val remaining: Int = 0)

@Serializable
data class ReferredBy(val status: String = "", val code: String = "")

/** The result of a redeem attempt: taken, or refused with a machine reason. */
data class RedeemResult(val ok: Boolean, val reason: String? = null)

/**
 * US-1385 (iOS `RedeemRejection`): the edge's `error_code` in plain words.
 *
 * Pure and separate from the network call, because the whole point of the
 * server tagging its rejections is that the client can say something SPECIFIC.
 * "That code isn't valid" in front of someone who typed their own code is a
 * dead end; "that's your own code" tells them what to do next.
 */
object RedeemRejection {

    @StringRes
    val GENERIC: Int = R.string.referral_error_generic

    @StringRes
    fun message(reason: String?): Int = when (reason) {
        "invalid_code", "not_found" -> R.string.referral_error_not_found
        "self_referral" -> R.string.referral_error_self
        "already_referred" -> R.string.referral_error_already
        "expired" -> R.string.referral_error_expired
        "account_suspended" -> R.string.referral_error_suspended
        "missing_code" -> R.string.referral_error_missing
        else -> GENERIC
    }
}

/**
 * The referral surface's derived numbers.
 *
 * Pure, so the arithmetic that a seller will check against their credit balance
 * is checkable here rather than only on a device.
 */
object Referrals {

    /** Where a shared link points. */
    const val SIGNUP_ORIGIN = "https://gradethread.com"

    fun link(code: String?): String? = code?.takeIf { it.isNotBlank() }?.let { "$SIGNUP_ORIGIN/signup?ref=$it" }

    /**
     * Referrals that are counted but not yet rewarded.
     *
     * Derived as total minus granted, NOT pending plus qualified (US-1255), so
     * the three columns always reconcile — Referred = In progress + Rewarded —
     * even if the server adds a fourth `reward_status` this build has never
     * heard of. Clamped at zero: a server that reports more granted than total
     * is wrong, and a negative count on screen is worse than a stale one.
     */
    fun inProgress(stats: ReferralStats): Int = maxOf(0, stats.total - stats.granted)

    fun alreadyReferred(me: ReferralMe?): Boolean = me?.referredBy != null

    fun canRedeem(me: ReferralMe?, typed: String, busy: Boolean): Boolean =
        !alreadyReferred(me) && !busy && typed.trim().isNotEmpty()

    /** Codes are stored uppercase; the edge uppercases too, so match it here. */
    fun normalize(typed: String): String = typed.trim().uppercase()

    /**
     * What the sharer actually sends: the code and the link.
     *
     * The code is in the message as well as in the link, because plenty of
     * places strip or shorten URLs and a code someone can type is the fallback.
     *
     * US-2976: the two PARTS, not the sentence.
     *
     * This is the text a seller sends a friend, so it has to be in the
     * seller's language - and the sentence around the code cannot be built by
     * an object with no Context. The screen formats
     * R.string.referral_share_text with these two.
     */
    fun shareParts(code: String?): Pair<String, String>? {
        val trimmed = code?.takeIf { it.isNotBlank() } ?: return null
        val url = link(trimmed) ?: return null
        return trimmed to url
    }

    /** What a referral is worth, or null when it is nothing. */
    fun creditsPerReferral(credits: ReferralCredits): Int? = credits.perReferral.takeIf { it > 0 }

    /** The next milestone's remaining count and bonus, or null when none is next. */
    fun nextMilestone(milestones: ReferralMilestones): Pair<Int, Int>? {
        val next = milestones.next ?: return null
        if (next.remaining <= 0) return null
        return next.remaining to next.bonus
    }

    /**
     * How the seller's own referred-by state reads.
     *
     * `pending` and `qualified` are both "not yet", but they are not the same
     * "not yet" — one is waiting on the friend, the other on us — so they get
     * different words.
     */
    @StringRes
    fun referredByLabel(referredBy: ReferredBy?): Int? = when (referredBy?.status) {
        null -> null
        "granted" -> R.string.referral_referred_granted
        "qualified" -> R.string.referral_referred_qualified
        else -> R.string.referral_referred_pending
    }
}
